hail/src/main/scala/is/hail/expr/ir/EmitStream.scala

```diff
@@ -8,9 +8,396 @@ import is.hail.expr.types.physical._
+    def apply(eos: Code[Ctrl], push: A => Code[Ctrl])(implicit ctx: EmitStreamContext): Source[A] = {
+      val outerPullJP = joinPoint()
+      var innerSource: Source[A] = null
+      val innerPullJP = joinPoint()
+      val hasBeenPulled = newLocal[Code[Boolean]]
+      val outerSource = outer(
+        eos = eos,
+        push = inner => {
+          innerSource = inner(
+            eos = outerPullJP(()),
+            push = push)
+          innerPullJP.define(_ => innerSource.pull)
+          Code(innerSource.setup, innerPullJP(()))
+        })
+      outerPullJP.define(_ => outerSource.pull)
+      Source[A](
+        setup0 = Code(hasBeenPulled := false, outerSource.setup0, innerSource.setup0),
+        close0 = Code(innerSource.close0, outerSource.close0),
+        setup = Code(hasBeenPulled := false, outerSource.setup),
+        close = Code(hasBeenPulled.load.mux(innerSource.close, Code._empty),
+                     outerSource.close),
+        pull = hasBeenPulled.load.mux(innerPullJP(()),
+                                      Code(hasBeenPulled := true, outerPullJP(()))))
+    }
+  }
+
+  def filter[A](stream: Stream[COption[A]]): Stream[A] = new Stream[A] {
+    def apply(eos: Code[Ctrl], push: A => Code[Ctrl])(implicit ctx: EmitStreamContext): Source[A] = {
+      val pullJP = joinPoint()
+      val source = stream(
+        eos = eos,
+        push = _.apply(none = pullJP(()), some = push))
+      pullJP.define(_ => source.pull)
+      source.copy(pull = pullJP(()))
+    }
+  }
+
+  def filter[A: ParameterPack](stream: Stream[A], cond: A => Code[Boolean]): Stream[A] =
+    filter(mapCPS[A, COption[A]](stream)((_ctx, a, k) => {
+      implicit val ctx = _ctx
+      val as = newLocal[A]
+      Code(as := a, k(COption(!cond(as.load), as.load)))
+    }))
+
+  def zip[A, B](left: Stream[A], right: Stream[B]): Stream[(A, B)] = new Stream[(A, B)] {
+    def apply(eos: Code[Ctrl], push: ((A, B)) => Code[Ctrl])(implicit ctx: EmitStreamContext): Source[(A, B)] = {
+      val eosJP = joinPoint()
+      val leftEOSJP = joinPoint()
+      val rightEOSJP = joinPoint()
+      var rightSource: Source[B] = null
+      val leftSource = left(
+        eos = leftEOSJP(()),
+        push = a => {
+          rightSource = right(
+            eos = rightEOSJP(()),
+            push = b => push((a, b)))
+          rightSource.pull
+        })
+      leftEOSJP.define(_ => Code(rightSource.close, eosJP(())))
+      rightEOSJP.define(_ => Code(leftSource.close, eosJP(())))
+      eosJP.define(_ => eos)
+
+      Source[(A, B)](
+        setup0 = Code(leftSource.setup0, rightSource.setup0),
+        close0 = Code(leftSource.close0, rightSource.close0),
+        setup = Code(leftSource.setup, rightSource.setup),
+        close = Code(leftSource.close, rightSource.close),
+        pull = leftSource.pull)
+    }
+  }
+
+  def multiZip(streams: IndexedSeq[Stream[Code[_]]]): Stream[IndexedSeq[Code[_]]] = new Stream[IndexedSeq[Code[_]]] {
```

Comment:
Actually, in light of our discussion, why does his need to be Code[_] or PameterPack at all?  Can't you write

```
def multiZip[A](streams: IndexedSeq[Stream[A]]): Stream[IndexedSeq[A]] = ...
```

?

I like that.

Reply:
Ah, yeah, I think you're right. The binary zip doesn't assume `Code[_]`, multiZip shouldn't either.

Reply:
Oh, the problem is `multiZip[A]` implies all the streams you're zipping have the same element type.

Reply:
Is that a problem?  It works fine for Code[_], or some EmitWhatever.

Reply:
I see what you mean. Yeah, I can change that.

This did make me realize why we need both zip and multiZip.